flink-connectors/flink-connector-gcp-pubsub/src/main/java/org/apache/flink/streaming/connectors/gcp/pubsub/source/reader/PubSubSplitReader.java

```diff
@@ -0,0 +1,165 @@
+    private final PubSubSubscriber subscriber;
+    private final PubSubDeserializationSchema<T> deserializationSchema;
+    private final PubSubCollector collector;
+    // Store the IDs of GCP Pub/Sub messages that yet have to be acknowledged so that they are not
+    // resent.
+    private final List<String> messageIdsToAcknowledge = new ArrayList<>();
+
+    /**
+     * @param deserializationSchema a deserialization schema to apply to incoming message payloads.
+     * @param subscriber a subscriber object to read messages from.
+     */
+    public PubSubSplitReader(
+            PubSubDeserializationSchema deserializationSchema, PubSubSubscriber subscriber) {
+
+        this.subscriber = subscriber;
+        this.deserializationSchema = deserializationSchema;
+        this.collector = new PubSubCollector();
+    }
+
+    @Override
+    public RecordsWithSplitIds<Tuple2<T, Long>> fetch() throws IOException {
+        RecordsBySplits.Builder<Tuple2<T, Long>> recordsBySplits = new RecordsBySplits.Builder<>();
+
+        for (ReceivedMessage receivedMessage : subscriber.pull()) {
+            try {
+                // Deserialize messages into a collector so that logic in the user-provided
+                // deserialization schema decides how to map GCP Pub/Sub messages to records in
+                // Flink. This allows e.g. batching together multiple Flink records in a single GCP
+                // Pub/Sub message.
+                deserializationSchema.deserialize(receivedMessage.getMessage(), collector);
+                collector
+                        .getMessages()
+                        .forEach(
+                                message ->
+                                        recordsBySplits.add(
+                                                PubSubSplit.SPLIT_ID,
+                                                new Tuple2<>(
+                                                        message,
+                                                        // A timestamp provided by GCP Pub/Sub
+                                                        // indicating when the message was initially
+                                                        // published
+                                                        receivedMessage
+                                                                .getMessage()
+                                                                .getPublishTime()
+                                                                .getSeconds())));
+            } catch (Exception e) {
+                throw new IOException("Failed to deserialize received message due to", e);
+            } finally {
+                collector.reset();
+            }
+
+            messageIdsToAcknowledge.add(receivedMessage.getAckId());
+        }
+
+        if (collector.isEndOfStreamSignalled()) {
+            recordsBySplits.addFinishedSplit(PubSubSplit.SPLIT_ID);
+        }
```

Comment:
Does this fully work? If not I'd propose we remove the `isEndOfStreamSignalled` handling completely and simplifying the deserializer.

Reply:
Nope, it doesn't, I will remove the end-of-stream logic.